packages/amd-loader/src/amd-loader.ts

```diff
@@ -0,0 +1,237 @@
+type modCallback = (...args: {}[]) => void;
+
+interface Window {
+  define: (deps: string[], factory: modCallback) => void;
+}
+
+(function() {
+
+interface Module {
+  url: normalizedUrl;
+  urlBase: normalizedUrl;
+  exports: {[id: string]: {}};
+  /** This module hasn't started loading its script yet. */
+  needsLoad: boolean;
+  /** All dependencies are resolved and the factory has run. */
+  resolved: boolean;
+  /** Callbacks from dependents waiting for this module to resolve. */
+  notify: Array<() => void>;
+}
+
+/**
+ * A global map from a fully qualified module URLs to module objects.
+ */
+const registry: {[url: string]: Module} = Object.create(null);
+
+/**
+ * Return a module object from the registry for the given URL, creating one if
+ * it doesn't exist yet.
+ */
+function getModule(url: normalizedUrl): Module {
+  let mod = registry[url];
+  if (mod === undefined) {
+    mod = registry[url] = {
+      url,
+      urlBase: getUrlBase(url),
+      exports: Object.create(null),
+      resolved: false,
+      needsLoad: true,
+      notify: [],
+    };
+  }
+  return mod;
+}
+
+type normalizedUrl = string&{_normalized: never};
+
+const anchor = document.createElement('a');
+
+/**
+ * Use the browser to resolve a URL to its canonical format.
+ *
+ * Examples:
+ *
+ *  - //example.com/ => http://example.com/
+ *  - http://example.com => http://example.com/
+ *  - http://example.com/foo/bar/../baz => http://example.com/foo/baz
+ */
+function normalizeUrl(url: string): normalizedUrl {
+  anchor.href = url;
+  return anchor.href as normalizedUrl;
+}
+
+/**
+ * Examples:
+ *
+ *  - http://example.com/ => http://example.com/
+ *  - http://example.com/foo.js => http://example.com/
+ *  - http://example.com/foo/ => http://example.com/foo/
+ *  - http://example.com/foo/?qu/ery#fr/ag => http://example.com/foo/
+ */
+function getUrlBase(url: normalizedUrl): normalizedUrl {
+  url = url.split('?')[0] as normalizedUrl;
```

Comment:
can you use the anchor for this?

Reply:
`.pathname` could be useful, but then we need to reconstruct scheme, host, etc., so we think this is simpler for now.